### PR TITLE
Symfony flex support

### PR DESCRIPTION
**Pull request by zorn-v**
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes
| BC breaks?    | No
| Deprecations? | No
| Fixed tickets | #1437

### CODE REVIEW

File: recipe/symfony-flex.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Symfony Flex Configuration
+ */
+
+//No need to clear anything
+set('clear_paths', []);
+
+// Symfony shared dirs
+set('shared_dirs', ['var/log', 'var/sessions']);
```
Comment (Cryde) [inline recipe/symfony-flex.php]: same here

File: recipe/symfony-flex.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+// Symfony writable dirs
+set('writable_dirs', ['var/cache', 'var/log', 'var/sessions']);
```
Comment (Cryde) [inline recipe/symfony-flex.php]: same here

File: recipe/symfony-flex.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+// Symfony web dir
+set('web_dir', 'public');
+
+// Assets
+set('assets', ['public/css', 'public/images', 'public/js']);
+
+// Environment vars
+set('env', function () {
+    return [
+        'APP_ENV' => get('symfony_env')
```
Comment (Cryde) [inline recipe/symfony-flex.php]: We should probably add documentation to tell the user to put "symfony_env" in it's host.yml

File: recipe/symfony-flex.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+namespace Deployer;
+
+require_once __DIR__ . '/symfony3.php';
+
+/**
+ * Symfony Flex Configuration
+ */
+
+//No need to clear anything
+set('clear_paths', []);
```
Comment (Cryde) [inline recipe/symfony-flex.php]: It's already define in symfony3 recipe :)

File: recipe/symfony-flex.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+// Symfony shared dirs
+set('shared_dirs', ['var/log', 'var/sessions']);
+
+// Symfony writable dirs
+set('writable_dirs', ['var/cache', 'var/log', 'var/sessions']);
+
+//File for DotEnv if not using env vars
+set('shared_files', ['.env']);
```
Comment (Cryde) [inline recipe/symfony-flex.php]: And btw the dotenv component is in require-dev when we do a fresh install ... I'm not sure if it's a good idea

**Comment by Cryde** [comment]
Another tought : what about commands launched that require DB access (in prod mode) ?  🤔    
Like migration for instance

File: recipe/symfony-flex.php (hunk 5 above)
Comment (antonmedv) [reply recipe/symfony-flex.php]: Actually there is a conversation on moving dotenv to require https://github.com/symfony/symfony/issues/25643

File: recipe/symfony-flex.php (hunk 4 above)
Comment (zorn-v) [reply recipe/symfony-flex.php]: Nope

File: recipe/symfony-flex.php (hunk 1 above)
Comment (zorn-v) [reply recipe/symfony-flex.php]: Also nope. And 'var/logs', not 'var/log'

File: recipe/symfony-flex.php (hunk 2 above)
Comment (zorn-v) [reply recipe/symfony-flex.php]: Also `var/log`

File: recipe/symfony-flex.php (hunk 3 above)
Comment (zorn-v) [reply recipe/symfony-flex.php]: Before it was `SYMFONY_ENV` https://github.com/deployphp/deployer/blob/master/recipe/symfony.php#L41
But it not for flex

**Comment by zorn-v** [comment]
And now think about is it ENV vars really good for deploy ?

**Comment by antonmedv** [comment]
@zorn-v no, then don't. 😄 

I really want to combine all symfony recipes into one in next v7. Let's merge it now for v6 although. 

Can you add init script for symfony flex?

File: recipe/symfony-flex.php (hunk 4 above)
Comment (Cryde) [reply recipe/symfony-flex.php]: My bad ! Sorry !

**Comment by zorn-v** [comment]
What do you mean by 'init script' ?

**Comment by antonmedv** [comment]
https://github.com/deployphp/deployer/tree/master/src/Initializer/Template

**Comment by zorn-v** [comment]
Yes, I already understand what do you mean 😄 
How do you think, do it with accounting of DotEnv in prod ?

"Best practice" (env vars) not so conveniently... You need to set params in 3 different places. 
1. In deployer
2. For console commands on remote
3. For web server on remote

**Comment by antonmedv** [comment]
Let's leave it for user? 😁 

Trim every thing related of setting params from recipe and add comment to init template about for to start using `.env`, if user want of cause :))

I like `.env` file.

**Comment by zorn-v** [comment]
What do you think ?
Yes there need some workaround for `composer install` (coz scripts in composer.json)

File: src/Initializer/Template/SymfonyFlexTemplate.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    protected function getExtraContent()
+    {
+        return <<<PHP
+
+// Migrate database before symlink new release.
+
+before('deploy:symlink', 'database:migrate');
+
+//If you don't use DotEnv in prod just wipe it out
+task('clear:env', function () {
```
Comment (antonmedv) [inline src/Initializer/Template/SymfonyFlexTemplate.php]: What purpose of this task?

Comment (zorn-v) [reply src/Initializer/Template/SymfonyFlexTemplate.php]: DotEnv does not run at all if APP_ENV is set https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/3.3/public/index.php#L11

Comment (antonmedv) [reply src/Initializer/Template/SymfonyFlexTemplate.php]: I think it's better not to hack APP_ENV but unstead write comment what user need to move dotenv to require section.

Comment (zorn-v) [reply src/Initializer/Template/SymfonyFlexTemplate.php]: But this task needed exactly in order if DotEnv in require ;)
One more time - DotEnv does not run if APP_ENV is set.
If you don't set it in env, `composer install` will fail.

Yes I know that is hack. But I think that I can not convince community to change `!isset($_SERVER['APP_ENV'])` to `!file_exists(__DIR__.'/../.env')`

File: recipe/symfony-flex.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+// Symfony web dir
+set('web_dir', 'public');
+
+// Assets
+set('assets', ['public/css', 'public/images', 'public/js']);
```
Comment (gander) [inline recipe/symfony-flex.php]:
```php
set('assets', ['{{web_dir}}/css', '{{web_dir}}/images', '{{web_dir}}/js']);
```

File: recipe/symfony-flex.php (hunk 3 above)
Comment (gander) [reply recipe/symfony-flex.php]: Maybe:
```php
// Environment vars
set('env', function () {
    return [
        'SYMFONY_ENV' => get('symfony_env'),
        'APP_ENV'     => get('symfony_env'),
    ];
});
```

File: recipe/symfony-flex.php (hunk 1 above)
Comment (LouTerrailloune) [reply recipe/symfony-flex.php]: The default directory for logs is `var/log`  (since 3.4)

File: recipe/symfony-flex.php (hunk 3 above)
Comment (gander) [reply recipe/symfony-flex.php]: I can not set environment variables on my server so `Dotenv` is also included in the _production_ environment. There were complications. I noticed that setting `APP_ENV` disables reading the _.env_ file, therefore I set only `symfony_env` to have all commands with `--env = {{symfony_env}}` option

Comment (zorn-v) [reply recipe/symfony-flex.php]: Hmm. Maybe you are right and your solution is much better than my hacky workaround https://github.com/deployphp/deployer/pull/1440#discussion_r159818610
AFAIR composer does not account `--env=`
Will test, and change it if all works fine.

**Comment by antonmedv** [comment]
Okay, problem now with:
```
if (!isset($_SERVER['APP_ENV'])) {
    if (!class_exists(Dotenv::class)) {
        throw new \RuntimeException('APP_ENV environment variable is not defined. You need to define environment variables for configuration or add "symfony/dotenv" as a Composer dependency to load variables from a .env file.');
    }
    (new Dotenv())->load(__DIR__.'/../.env');
}
```

I think we just need mention not to refactor to:

```
(new Dotenv())->load(__DIR__.'/../.env');
```

**Comment by Cryde** [comment]
Any news on this ?

**Comment by antonmedv** [comment]
Need to figure out how to properly handle env. Soon will create my recipe.

**Comment by zorn-v** [comment]
Any news, suggestions. Maybe need some suggestions ?

**Comment by zorn-v** [comment]
It last project I just installed apache-pack and put SetEnv in .htaccess
ENV for deployer only needed for doctrine migrations (except APP_ENV=prod). Maybe think in this direction ?

**Comment by antonmedv** [comment]
I’m thinking on suggesting of using doyens for prod. This will require APP_ENV rrmoval

**Comment by antonmedv** [comment]
Done in 443de26e87d1c1f3185ebf445eb625bacaec23d6

**Comment by zorn-v** [comment]
Maybe `var/sessions` also in shared ?

**Comment by antonmedv** [comment]
`var` stands for variable?

**Comment by antonmedv** [comment]
`var/sessions` yes, my bad. Should go to shared.

**Comment by zorn-v** [comment]
About `writable_mode` - I get errors with `acl` one. `chown` is my bet )

**Comment by antonmedv** [comment]
But you should set `writable_mode` in deploy.php

**Comment by zorn-v** [comment]
Yep, but I talk about default one.

**Comment by antonmedv** [comment]
No, but chown may require sudo. acl no. And acl more flexible. 

But I'm thinking to refactor it in v7.

**Comment by Cryde** [comment]
Will there be a 6.0.1 ? for the missing ```var/sessions``` ? :p

**Comment by antonmedv** [comment]
Yes :) 
Sorry for delay.

**Comment by Cryde** [comment]
Don't be sorry ! You do an amazing work ! We can still add it ourself :D

**Pull request closed without merging**